core/TypeConstraint.cc

```diff
@@ -276,4 +276,41 @@ std::string TypeConstraint::toString(const core::GlobalState &gs) const {
     return to_string(buf);
 }
 
+vector<ErrorLine> TypeConstraint::toExplanation(const core::GlobalState &gs) const {
+    auto boundsFor = UnorderedMap<SymbolRef, pair<TypePtr, TypePtr>>{};
+
+    for (const auto &[sym, lowerBound] : this->lowerBounds) {
+        auto &[lowerRef, _upperRef] = boundsFor[sym];
+        ENFORCE(lowerRef == nullptr, "{} in lowerBounds twice?", sym.show(gs));
+        lowerRef = lowerBound;
+    }
+    for (const auto &[sym, upperBound] : this->upperBounds) {
+        auto &[_lowerRef, upperRef] = boundsFor[sym];
+        ENFORCE(upperRef == nullptr, "{} in upperBounds twice?", sym.show(gs));
+        upperRef = upperBound;
+    }
```

Comment:
You added this for pretty-printing the constraint set as well, what do you think about extracting it as a utility function?

Reply:
https://github.com/sorbet/sorbet/blob/c07eeaa6989dc16ae18ccfa7b0cf9852574f26f5/core/TypeConstraint.cc#L247-L262